### PR TITLE
[JSONSelection] Track location information for parsed structures using `Ranged<T>` trait and `WithRange<T>` wrapper

### PR DESCRIPTION
**Pull request by benjamn**
In order to provide more helpful parsing and error messages for JSONSelection, and to enable IDE/LSP features like syntax highlighting and tooltip hovering in the future, the JSONSelection parser needs to track where each parsed substructure came from within the original source string.

The [`nom_locate`](https://docs.rs/nom_locate/latest/nom_locate/) crate was helpful for replacing raw `&str` inputs with the `type Span<'a> = LocatedSpan<&'a str>` type, which `nom` (mostly) seamlessly accepts everywhere `&str` was previously expected. However, while the `LocatedSpan` type makes it easy to determine the _starting_ offset of a given AST structure, the _ending_ offset typically needs to be derived from parsing the structure's children. To bootstrap this parsing from the bottom up, I found it helpful to add a `parsed_span` combinator to match specific concrete tokens and capture their range information.

This would all have been considerably easier if every AST structure was a `struct`, because then I could simply have added a `range: Option<(usize, usize)>` (aka `range: OffsetRange`) field to each struct, and then make sure that field gets populated during parsing. Where possible, that's what I did, but there are a number of AST structures that make more sense as `enum` types, and we also need to keep track of location information for primitive/string types that cannot store their own ranges.

To work with this variety of AST structures, I introduced a `Ranged<T>` trait, which is implemented by all the structures that can provide (or cheaply derive) their own ranges. For those that cannot, there's a `WithRange<T>` wrapper struct that implements `Ranged<T>`. I considered using `WithRange<T>` (note: fka `Parsed<T>`) for everything, but that proved cumbersome and redundant, so the later commits in this PR walk back the use of `WithRange<T>` to just the types that need it, leaving the rest to implement `Ranged<T>` directly, without a wrapper.

To make this `range` information immediately useful, I am now including it as one of the fields of the `ApplyToError` struct, making runtime `apply_to_path` errors easier to debug. I would also like to include range information (along with more descriptive error messages) in parse errors generated by `nom`, but I haven't found a good way to replace the opaque `nom::error::Error` objects (which only store the input string and an error code) with errors that allow custom messages and range information, so that will have to wait for a subsequent PR.

You can review this PR commit-by-commit if you like, but I changed my mind a few times (e.g. renaming `Parsed<T>` to `WithRange<T>` after adding `Range<T>`), so it might be easier to review the whole set of changes together.

### CODE REVIEW

**Comment by benjamn** [comment]
[Those test failures](https://app.circleci.com/pipelines/github/apollographql/router/25977/workflows/a0cd18eb-5416-4888-8eef-2a34c32c5869/jobs/187432) are legit, since I neglected to update snapshots in the `apollo-router` crate. I'll take the PR back out of draft once I have that fixed.

**Review by lennyburdette — APPROVED** [review]
i didn't go over this with a fine-toothed comb, but the overall structure looks sound and the tests look good. i'm inclined to merge this early, probably right after our next release, so we can iterate on diagnostics for the language server. thank you for this!

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (nicholascioli) [inline apollo-federation/src/sources/connect/json_selection/location.rs:23]: Is there a reason this is encoded as a tuple versus the [built-in range type](https://doc.rust-lang.org/std/ops/struct.Range.html)?

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (nicholascioli) [inline apollo-federation/src/sources/connect/json_selection/location.rs:103]: What if ranges overlap or aren't in increasing order? Might be useful to have a guard here:

```suggestion
        (Some((left_start, _left_end)), Some((right_start, right_end))) if left_start <  right_start => {
            Some((left_start, right_end))
        }
```

Or something similar to ensure that this range is never invalid

**Review by nicholascioli — CHANGES_REQUESTED** [review]
First pass! I'll need to continue looking over the rest tomorrow.

File: apollo-federation/src/sources/connect/json_selection/apply_to.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/apply_to.rs:119]: Huh, I think my comment was mistaken about `serde_json[_bytes]::Value` not implementing `Hash`. I will definitely switch to the auto-derived version.

As for not including `self.range`, I think I'm generally coming around to the reality that while two `ApplyToError`s may have the same error message, the `error.path` already basically ensures they will never/rarely be subject to deduplication, so it's not a problem that `error.range` truly eliminates any opportunity for deduplication.

Of course, that raises the question of why we're using an `IndexSet` to accumulate the errors internally, which we've previously discussed. I'll address that in a follow-up commit/PR, if that's OK.

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs:23]: That's a great idea!

File: apollo-federation/src/sources/connect/json_selection/README.md
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/README.md:386]: Right, see https://github.com/apollographql/router/pull/5987/commits/0314539954d4aa5bbf1af61c0e03478bff0804ed. I included that change in this PR because I believed it would affect how `Alias` tracks its location (by wrapping a `Key`). I still think it's a worthwhile change, though I suppose we could have used `WithRange<String>` to achieve a similar effect, without switching to `Key`.

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs]: https://github.com/apollographql/router/pull/5987/commits/774ae142dcb59a8699bf1c685a2d48c00bf56311

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs]: Included in https://github.com/apollographql/router/pull/5987/commits/774ae142dcb59a8699bf1c685a2d48c00bf56311.

File: apollo-federation/src/sources/connect/json_selection/apply_to.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/apply_to.rs]: https://github.com/apollographql/router/pull/5987/commits/d9b1513eb39ec6baeb23e805211b475acda3185e

File: apollo-federation/src/sources/connect/json_selection/apply_to.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/apply_to.rs]: https://github.com/apollographql/router/pull/5987/commits/fa760db2130adca820777597cf5c8f86812fe604

File: apollo-federation/src/sources/connect/json_selection/apply_to.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/apply_to.rs]: https://github.com/apollographql/router/pull/5987/commits/f599aed8b59e5defb84c3a5038c9383391144df9

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (nicholascioli) [reply apollo-federation/src/sources/connect/json_selection/parser.rs:146]: You could even add a trait bound to the trait itself to catch any instances where deref isn't implemented. Something like:

```rust
pub trait Ranged<T>: Deref<Target = T> {
  ...
}
```

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (nicholascioli) [inline apollo-federation/src/sources/connect/json_selection/parser.rs:1]: It might be nice if most of the beefier tests could use snapshots to make updating the tests later a bit easier.

File: apollo-federation/src/sources/connect/json_selection/pretty.rs
Comment (nicholascioli) [inline apollo-federation/src/sources/connect/json_selection/pretty.rs:1]: You might not need to change the impl blocks of all of the AST nodes since you've implemented Deref, which should know how to delegate the call to the dereferenced inner type.

**Review by nicholascioli — CHANGES_REQUESTED** [review]
Second half done! I'd still like someone else to do a quick look-over, but once these changes get fixed it should be good to merge!

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs:127]: I certainly thought about that, but I feel like the `#[cfg(test)]` conveys it's test-only, and I wanted to be able to import the `StripRanges` trait in `parser.rs` and `lit_expr.rs` tests without having to import the `location.rs` tests module.

File: apollo-federation/src/sources/connect/json_selection/methods/tests.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/methods/tests.rs:693]: Yes, I wanted to test that the end offset of the `$->get($->echo(-5)->mul(2)` expression does not include the trailing spaces.

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/parser.rs:146]: I'm implementing `Ranged<T>` for some types `T` that do not implement `Deref`, whose `node` methods return `self` (like `PathSelection` and `NamedSelection`), so I don't think that trait bound will work.

Something I could do, instead of requiring this `node` method, is to rely on the `AsRef` trait implemented for `WithRange<T>`, and just handle the non-`WithRange` implementations of `Ranged<T>` as usual (no need to call `.as_ref()`. There's already some ambiguity about when I use `wrapped.node()` vs. `wrapped.as_ref()`, so I suspect we can do without `node`.

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/parser.rs:1]: I'm not a big fan of external snapshot tests, though they are easy to update, because in my experience they tend to get updated blindly without much review by folks (including the author) just trying to make (legitimately failing) tests pass, but I agree the larger tests are annoying to update, so it's a balance.

File: apollo-federation/src/sources/connect/json_selection/pretty.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/pretty.rs:1]: That was my hope too, and `Deref` definitely helps (especially in the `apply_to_path` code), but I implemented `Deref` early and then fixed everything that was still broken, so unfortunately I don't think it covered all cases.

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs]: I'll switch this one out, since I can imagine it being pretty annoying to keep updated.

File: apollo-federation/src/sources/connect/json_selection/methods/tests.rs
Comment (nicholascioli) [reply apollo-federation/src/sources/connect/json_selection/methods/tests.rs:693]: Maybe add a comment so that it doesn't get removed accidentally later! I could see myself forgetting and accidentally removing it later.

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (nicholascioli) [reply apollo-federation/src/sources/connect/json_selection/parser.rs:1]: I'm mostly worried that we'll encounter an issue with the file length later as experienced by @dylan-apollo since we're locked at the current version of rust until we remove deno, but I'll defer to their judgement.

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs]: https://github.com/apollographql/router/pull/5987/commits/eca341b243fa98ce826a3e033a3b3e483344af8b

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/parser.rs:1]: https://github.com/apollographql/router/pull/5987/commits/eca341b243fa98ce826a3e033a3b3e483344af8b

File: apollo-federation/src/sources/connect/json_selection/parser.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/parser.rs]: Successfully removed `Ranged<T>::node()`: https://github.com/apollographql/router/pull/5987/commits/4005495847798865a6544f62f4dc7a0898874ec7

File: apollo-federation/src/sources/connect/json_selection/pretty.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/pretty.rs:1]: Oh, sorry, GitHub wasn't showing me the code you were referring to, but I think I understand what you're suggesting now:

Screenshot 2024-09-12 at 3 59 23 PM

You're saying the `impl PrettyPrintable for WithRange<T>` lines could have stayed `impl PrettyPrintable for T`?

Comment (nicholascioli) [reply apollo-federation/src/sources/connect/json_selection/pretty.rs:1]: Yes! Sorry, I may have left the comment on an outdated view of the code.

Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/pretty.rs:1]: Ok, yes, that worked: https://github.com/apollographql/router/pull/5987/commits/1454e5967d6e16438a183f41c71c5efc35e8ebe0

For what it's worth, I believe I did need to switch the `impl ApplyToInternal for T` lines to `impl ApplyToInternal for WithRange<T>`, because that way `self` refers to the `WithRange` wrapper, so the `apply_to_path` methods have access to the location information for the applied type. But `PrettyPrintable` doesn't need that location information, so the same change is not necessary there, like you pointed out.

File: apollo-federation/src/sources/connect/json_selection/apply_to.rs
Comment (nicholascioli) [reply apollo-federation/src/sources/connect/json_selection/apply_to.rs]: Just to make sure, you expect the range to be along the mathematical range `[min, max)`, right? If you need it to include max, then change the range to min..=max.

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (nicholascioli) [reply apollo-federation/src/sources/connect/json_selection/location.rs]: What should happen if the ranges aren't continuous e.g. 1..3 and 9..10. Should they be merged to 1..10?

File: apollo-federation/src/sources/connect/json_selection/apply_to.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/apply_to.rs]: Yep, `[min, max)` is my goal (like the `->slice(start, end)` method).

File: apollo-federation/src/sources/connect/json_selection/location.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/location.rs]: Yes, often the children of a given AST node will be separated by spaces/comments, so they won't be contiguous, but the parent range should encompass all the child ranges.

**Review by nicholascioli — APPROVED** [review]

**Comment by nicholascioli** [comment]
Thanks for working through all of the comments! The only thing left is if you want to add a quick comment about the trailing space. Otherwise, feel free to squash and merge!

File: apollo-federation/src/sources/connect/json_selection/methods/tests.rs
Comment (benjamn) [reply apollo-federation/src/sources/connect/json_selection/methods/tests.rs]: https://github.com/apollographql/router/pull/5987/commits/b08dbb958d2c290ec90f4e2673b6e17af7c9b8f5

**Pull request merged**
